test/jdk/java/util/zip/ZipInputStream/ZipInputStreamCRCCheck.java

```diff
@@ -0,0 +1,162 @@
+import java.util.List;
+import java.util.zip.CRC32;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+/*
+ * @test
+ * @bug 8354799
+ * @summary verifies that when using java.util.zip.ZipInputStream any incorrect CRC values for
+ *          entries results in an exception when reading those entries
+ * @run junit ZipInputStreamCRCCheck
+ */
+public class ZipInputStreamCRCCheck {
+
+    static List<Arguments> args() {
+        final int[] entrySizes = {10, 0, 1024};
+        final int[] methods = new int[]{ZipOutputStream.STORED, ZipOutputStream.DEFLATED};
+        final List<Arguments> args = new ArrayList<>();
+        for (int method : methods) {
+            for (int size : entrySizes) {
+                args.add(Arguments.of(method, size));
+            }
+        }
+        return args;
+    }
+
+    /*
+     * Creates a ZIP file containing an entry whose CRC value is set to an incorrect
+     * value. The test then uses java.util.zip.ZipInputStream to read that ZIP file
+     * and verifies that reading the entry's contents results in a ZipException due
+     * to the incorrect CRC value.
+     */
+    @ParameterizedTest
+    @MethodSource("args")
+    void test(final int entryMethod, final int entrySize) throws Exception {
+        final String entryName = entryMethod == ZipOutputStream.STORED
+                ? "stored-entry-size=" + entrySize
+                : "deflated-entry-size=" + entrySize;
+        System.out.println("creating ZIP with entry " + entryName);
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try (final ZipOutputStream zos = new ZipOutputStream(baos)) {
+
+            final ZipEntry entry = new ZipEntry(entryName);
+            entry.setSize(entrySize);
+            entry.setMethod(entryMethod);
+
+            final byte[] entryContent = new byte[entrySize];
+            final CRC32 crc = new CRC32();
+            crc.update(entryContent);
+            entry.setCrc(crc.getValue());
+
+            zos.putNextEntry(entry);
+            zos.write(entryContent);
+            zos.closeEntry();
+        }
+        final byte[] zipContent = baos.toByteArray();
+        // now change the CRC value of the entry to an incorrect value
+        changeEntryCRCValue(zipContent);
+        // use ZipInputStream to read the entry from that ZIP containing the incorrect CRC value
+        try (final ZipInputStream zis = new ZipInputStream(new ByteArrayInputStream(zipContent))) {
+            final ZipEntry entry = zis.getNextEntry();
+            assertEquals(entryName, entry.getName(), "unexpected entry name");
+            try {
```

Comment:
I would consider changing this to something such as we did in open/test/jdk/java/util/zip/ZipInputStream/Zip64DataDescriptor.java:

` ZipException ex = assertThrows(ZipException.class, () -> {
            readZipInputStream(invalidZip64);
        });
        String msg = String.format("Expected exeption message to contain 'invalid entry size', was %s",
                ex.getMessage());
        assertTrue(ex.getMessage().contains("invalid entry size"), msg);
    }`